bigflow/log.py

```diff
@@ -1,89 +1,136 @@
+        )
+
+    def create_logging_client(self):
+        return logging_v2.LoggingServiceV2Client()
+
+    def emit(self, record: logging.LogRecord):
+        cl_log_level = record.levelname  # CloudLogging list of supported log levels is a superset of python logging level names
+        json = {
+            'message': record.getMessage(),
+            'name': record.name,
+            'filename': record.filename,
+            'module': record.module,
+            'lineno': record.lineno,
         }
 
-    def warning(self, message):
-        self.write_log_entries(message, 'WARNING')
+        if record.exc_info and not record.exc_text:
+            # mimic caching behaviour of `logging.Formatter.format`
+            record.exc_text = self._format_exception(record.exc_info)
+        if record.exc_text:
+            json['exc_text'] = str(record.exc_text)
 
-    def error(self, message):
-        self.write_log_entries(message, 'ERROR')
+        if record.stack_info:
+            json['stack_info'] = str(record.stack_info)
 
-    def info(self, message):
-        self.write_log_entries(message, 'INFO')
+        self.write_log_entries(json, cl_log_level)
 
-    def write_log_entries(self, message, severity):
-        log_name = f"projects/{self.project_id}/logs/{self.logger_name}"
-        labels = {"id": f"{self.workflow_id or self.project_id}"}
-        entry = logging_v2.types.LogEntry(
-            log_name=log_name,
-            resource=self.get_resource(),
-            text_payload=message,
-            severity=severity,
-            labels=labels
-        )
-        self.client.write_log_entries([entry])
+    def _format_exception(self, ei):
+        etype, value, tb = ei
+        return "\n".join(traceback.format_exception(etype, value, tb))
 
-    def get_gcp_logs_message(self):
-        query = quote_plus(f'''logName="projects/{self.project_id}/logs/{self.logger_name}"
-labels.id="{self.workflow_id or self.project_id}"''')
-        return f'\n'\
-               f'*************************LOGS LINK************************* \n ' \
-               f'You can find this workflow logs here: https://console.cloud.google.com/logs/query;query={query} \n' \
-               f'***********************************************************'
+    def write_log_entries(self, json, severity):
+        entry = logging_v2.types.LogEntry()
+        entry.CopyFrom(self._log_entry_prototype) 
+        entry.json_payload.update(json)
+        entry.severity = LogSeverity[severity]
+        self.client.write_log_entries([entry])
 
 
-def handle_uncaught_exception(logger):
+def _uncaught_exception_handler(logger):
     def handler(exception_type, value, traceback):
         logger.error(f'Uncaught exception: {value}', exc_info=(exception_type, value, traceback))
     return handler
 
 
-class BigflowLogging(object):
-    IS_LOGGING_SET = False
+_LOGGING_CONFIGURED = False
+
+
+# TODO: move to some shared config module?
+class LogConfigDict(TypedDict):
```

Comment:
+1

Reply:
I'm going to do it later.

Reply:
I wanted to say that I like the TypedDict :)